x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plugin/SqlMediaTypeParser.java

```diff
@@ -0,0 +1,147 @@
+     * {@link XContent} outputs we can't use {@link RestToXContentListener}
+     * like everything else. We want to stick as closely as possible to
+     * Elasticsearch's defaults though, while still layering in ways to
+     * control the output more easily.
+     *
+     * First we find the string that the user used to specify the response
+     * format. If there is a {@code format} parameter we use that. If there
+     * isn't but there is a {@code Accept} header then we use that. If there
+     * isn't then we use the {@code Content-Type} header which is required.
+     */
+    public static SqlMediaType getResponseMediaType(RestRequest request, SqlQueryRequest sqlRequest) {
+        if (Mode.isDedicatedClient(sqlRequest.requestInfo().mode())
+            && (sqlRequest.binaryCommunication() == null || sqlRequest.binaryCommunication())) {
+            // enforce CBOR response for drivers and CLI (unless instructed differently through the config param)
+            return SqlMediaType.xContentType(XContentType.CBOR);
+        } else if (request.hasParam(URL_PARAM_FORMAT)) {
+            return validateColumnarRequest(sqlRequest.columnar(), mediaTypeFromParams(request), request);
+        }
+
+        return mediaTypeFromHeaders(request);
+    }
+
+    public static SqlMediaType getResponseMediaType(RestRequest request) {
+        return request.hasParam(URL_PARAM_FORMAT)
+            ? checkNonNullMediaType(mediaTypeFromParams(request), request)
+            : mediaTypeFromHeaders(request);
+    }
+
+    private static SqlMediaType mediaTypeFromHeaders(RestRequest request) {
+        String acceptType = getAcceptValue(request);
+        SqlMediaType mediaType = acceptType != null
+            ? SqlMediaType.fromMediaTypeOrFormat(acceptType)
+            : SqlMediaType.xContentType(request.getXContentType());
+        return checkNonNullMediaType(mediaType, request);
+    }
+
+    private static SqlMediaType mediaTypeFromParams(RestRequest request) {
+        return SqlMediaType.fromMediaTypeOrFormat(request.param(URL_PARAM_FORMAT));
+    }
+
+
+    private static SqlMediaType validateColumnarRequest(boolean requestIsColumnar, SqlMediaType fromMediaType, RestRequest request) {
+        if (requestIsColumnar && fromMediaType.isTextFormat()) {
+            throw new IllegalArgumentException("Invalid use of [columnar] argument: cannot be used in combination with "
+                + "txt, csv or tsv formats");
+        }
+        return checkNonNullMediaType(fromMediaType, request);
+    }
+
+    private static SqlMediaType checkNonNullMediaType(SqlMediaType mediaType, RestRequest request) {
+        if (mediaType == null) {
+            String msg = String.format(Locale.ROOT, "Invalid request content type: Accept=[%s], Content-Type=[%s], format=[%s]",
+                request.header("Accept"), request.header("Content-Type"), request.param(URL_PARAM_FORMAT));
+            throw new IllegalArgumentException(msg);
```

Comment:
Is there a more suitable `ElasticsearchStatusException` instead of the generic IAE?

Reply:
True, that might be more suitable, but master emits the IAE in this case, it might be better to keep it as is for consistency.